react/features/calendar-sync/web/microsoftCalendar.js

```diff
@@ -0,0 +1,564 @@
+/* @flow */
+
+import { Client } from '@microsoft/microsoft-graph-client';
+import rs from 'jsrsasign';
+
+import { createDeferred } from '../../../../modules/util/helpers';
+
+import { CALENDAR_TYPE } from '../constants';
+import {
+    setCalendarAPIAuthState,
+    setCalendarProfileEmail
+} from '../actions';
+
+/**
+ * Constants used for interacting with the Microsoft API.
+ *
+ * @private
+ * @type {object}
+ */
+const MS_API_CONFIGURATION = {
+    /**
+     * The URL to use when authenticating using Microsoft API.
+     * @type {string}
+     */
+    AUTH_ENDPOINT:
+        'https://login.microsoftonline.com/common/oauth2/v2.0/authorize?',
+
+    CALENDAR_ENDPOINT: '/me/calendars',
+
+    /**
+     * The Microsoft API scopes to request access for calendar.
+     *
+     * @type {string}
+     */
+    MS_API_SCOPES: 'openid profile Calendars.Read',
+
+    /**
+     * See https://docs.microsoft.com/en-us/azure/active-directory/develop/
+     * v2-oauth2-implicit-grant-flow#send-the-sign-in-request.
+     *
+     * FIXME BEFORE MERGE: Why is this needed?
+     *
+     * @type {string}
+     */
+    MS_CONSUMER_TENANT: '9188040d-6c67-4c5b-b112-36a304b66da',
+
+    /**
+     * The redirect URL to be used by the Microsoft API on successful
+     * authentication.
+     *
+     * @type {string}
+     */
+    REDIRECT_URI: `${window.location.origin}/static/msredirect.html`
+};
+
+/**
+ * Store the window from an auth request. That way it can be reused if a new
+ * request comes in and it can be used to indicate a request is in progress.
+ *
+ * @private
+ * @type {Object|null}
+ */
+let popupAuthWindow = null;
+
+export const microsoftCalendarApi = {
+    /**
+     * Retrieves the current calendar events.
+     *
+     * @param {number} fetchStartDays - The number of days to go back
+     * when fetching.
+     * @param {number} fetchEndDays - The number of days to fetch.
+     * @returns {function(Dispatch<*>): Promise<CalendarEntries>}
+     */
+    getCalendarEntries(fetchStartDays: ?number, fetchEndDays: ?number) {
+        return (dispatch: Dispatch<*>, getState: Function): Promise<*> => {
+            const state = getState()['features/calendar-sync'] || {};
+            const token = state.msAuthState && state.msAuthState.accessToken;
```

Comment:
Some token validation function should be used here maybe. Maybe it's in the token validation function that the token refreshing should happen.